### PR TITLE
Prevent static local function from referencing generic local function

### LINKED ISSUE #39706: Static local function should not be able to call non-static generic local function from parent scope
**Version Used**: VS2019 16.4.0 Preview3

Seems fix for #38129 not works for generic local function.

**Steps to Reproduce**:

https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKFygZgAJ0iBhAG2igyIG9cinmmUMA2EgFiIFk4YACwD2AEwAUASkYtmDHLMVEAlgDsYKogF4iABgDcMpcxQ8AMsIDGECgB4AKgD4pR4/Vdvmy7XsMLPRAC+HkpsaEQQPlLajioA1HF+ASQYAJziEJJJAWycpkQAyjAQMMqWFtYULv6e8sksFTa2ajDOWSGKwTVuRSVljVXt3SxdgUA===
```cs
        static void Method()
        {
            int i = 0;
            void Local<T>()
            {
                i = 0;
            }
            Action a = () => i++;
            Console.WriteLine(a);
            static void StaticLocal()
            {
                Local<int>();
            }
            StaticLocal();
        }
```

**Expected Behavior**:
It should not compile. And get an error as it did when used without generic.
`error CS8421: A static local function cannot contain a reference to 'Local'.`
See `Item 2`(#38129) in https://github.com/dotnet/roslyn/blob/master/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20post%20VS2019.md

**Actual Behavior**:
It compile without error, and get an non-static method.

### PR DESCRIPTION
**Pull request by cston**
Fixes #39706.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
Looking at the linked issue, I'm wondering if there's an assert we can add to ensure that, at least in debug compilers, the lowered method produced for a static local function is always static. Perhaps that would help us catch more of these kinds of corner cases?

**Comment by cston** [comment]
@RikkiGibson 
> I'm wondering if there's an assert we can add to ensure that, at least in debug compilers, the lowered method produced for a static local function is always static.

Good idea. I've added an assert to `SynthesizedClosureMethod`. Ideally the assert would be in emit that would require more changes since the core layer that is unaware of local functions.

File: src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs [hunk 1]
```diff
@@ -349,10 +356,7 @@ public override BoundNode VisitCall(BoundCall node)
         {
             VisitCall(node.Method, null, node.Arguments, node.ArgumentRefKindsOpt, node.ArgumentNamesOpt, node.Expanded, node);
             CheckReceiverIfField(node.ReceiverOpt);
-            if (node.Method is LocalFunctionSymbol)
-            {
-                CheckReferenceToVariable(node, node.Method);
-            }
+            CheckReferenceToMethodIfLocalFunction(node, node.Method);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs]: >node.Method [](start = 56, length = 11)

Is this guaranteed to be not null?

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs [hunk 2]
```diff
@@ -4947,6 +4999,85 @@ static void F2()
...
+            var source =
+@"#pragma warning disable 8321
+class Program
+{
+    static void Method()
+    {
+        void Local<T>() {}
+        static System.Action StaticLocal()
+        {
+            return Local<int>;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: >Local<int> [](start = 19, length = 10)

Consider also testing the form that explicitly invokes delegate constructor.

**Pull request merged**